File: CPP/1-Creationals/Object Pool/ObjectPool.cpp
#include <string>
#include <iostream>
#include <list>

class ImageResource
{
    std::string _url;

public:
    ImageResource() : _url("")
    {
    }

    void setImageURL(std::string url)
    {
        _url = url;
    }

    std::string getImageURL() const
    {
        return _url;
    }

    void reset()
    {
        _url = "";
    }
};

/*
Basic implementation of the ObjectPool.
Note, that this class is a singleton (see Singleton example for full implementation).
*/
class ObjectPool
{
private:
    std::list<ImageResource *> _images;

    ObjectPool() {}

public:
    /**
     * Static method for accessing class instance.
     * Part of Singleton design pattern.
     *
     * @return ObjectPool instance.
     */
    static ObjectPool &getInstance()
    {
        static ObjectPool instance;
        return instance;
    }

    /**
     * Returns instance of ImageResource.
     *
     * New ImageResource will be created if all the ImageResources
     * were used at the time of the request.
     *
     * @return ImageResource instance.
     */
    ImageResource* getImageResource()
    {
        if (_images.empty())
        {
            std::cout << "Creating new." << std::endl;
            return new ImageResource;
        }
        else
        {
            std::cout << "Reusing existing." << std::endl;
            ImageResource *img = _images.front();
            _images.pop_front();
            return img;
        }
    }

    /**
     * Return ImageResource back to the pool.
     *
     * The ImageResource must be initialized back to
     * the default settings before someone else
     * attempts to use it.
     *
     * @param object ImageResource instance.
     * @return void
     */
    void returnImageResource(ImageResource *object)
    {
        object->reset();
        _images.push_back(object);
    }

    void freePool()
    {
        // free memory for the objects that still are present in the pool
        while(_images.size())
        {
            ImageResource *obj = _images.front();
            _images.pop_front();
            delete obj;
        }
    }
};


int main()
{
    ObjectPool pool = ObjectPool::getInstance();
    ImageResource *one;
    ImageResource *two;

    /* ConnectionResource will be created. */
    one = pool.getImageResource();
    one->setImageURL("https://www.dataresources.com/server_entry_point.php");
    std::cout << "one = " << one->getImageURL() << " [" << one << "]" << std::endl;

    two = pool.getImageResource();
    two->setImageURL("https://www.another-dataresources.com/another_server_entry_point.php");
    std::cout << "two = " << two->getImageURL() << " [" << two << "]" << std::endl;

    // Return the connection resource to the pool because we don't need it anymore.
    pool.returnImageResource(one);
    pool.returnImageResource(two);

    /* ImageResources will be reused.
     * Notice that the ImageURL of both resources were reset back to zero.
     */
    one = pool.getImageResource();
    std::cout << "one = " << one->getImageURL() << " [" << one << "]" << std::endl;

    two = pool.getImageResource();
    std::cout << "two = " << two->getImageURL() << " [" << two << "]" << std::endl;

    pool.freePool();

    return 0;
}

//-----------------------------------------------------------
/*
Output:

Creating new.
one = https://www.dataresources.com/server_entry_point.php [0x21c6030]
Creating new.
two = https://www.another-dataresources.com/another_server_entry_point.php [0x21c60e0]
Reusing existing.
one =  [0x21c6030]
Reusing existing.
two =  [0x21c60e0]

*/

File: Python/1-Creationals/Object Pool/ObjectPool.py
#!/usr/bin/env python3


class ImageResource:
    """
    Test class used to demonstrate the ObjectPool pattern
    """
    def __init__(self):
        self.url = ''

    def reset(self):
        self.url = ''


_pool = None


class ObjectPool:
    """
    Singleton class representing the pool of ImageResource objects
    """
    def __init__(self):
        self.images = []

    def get_image_resource(self):
        if len(self.images) == 0:
            print("Creating new resource...")
            return ImageResource()
        else:
            print("Reusing existing resource...")
            return self.images.pop(0)

    def return_image_resource(self, resource):
        """
        Reset the given resource object and make it available for others that need it
        :param resource: Is the object released by the client code that needs to be reinserted into the pool
        :return: None
        """
        resource.reset()
        self.images.append(resource)

    def free_pool(self):
        """
        Remove all the objects from the pool
        :return: None
        """
        self.images.clear()


def shared_pool():
    """
    It checks that the instance of ObjectPool has been created or not and returns it to the caller
    :return: the unique instance of the ObjectPool
    """
    global _pool
    if _pool is None:
        _pool = ObjectPool()

    return _pool


def main():
    # Get the pool object
    pool = shared_pool()

    # Ask to the pool an instance of ImageResource
    one = pool.get_image_resource()
    # Customize it as needed by the client
    one.url = "https://www.dataresources.com/server_entry_point.php"
    print(one.url)

    # Ask another resource object to the pool
    two = pool.get_image_resource()
    # ... and customize it
    two.url = "https://www.another-dataresources.com/another_server_entry_point.php"
    print(two.url)

    # Return the resources to the pool because we don't need them anymore
    pool.return_image_resource(one)
    pool.return_image_resource(two)

    # Ask again the released resource
    one = pool.get_image_resource()
    # it has been reset by the pool, so it is at an initial state
    print(one.url)

    # same thing for thew second resource
    two = pool.get_image_resource()
    print(two.url)

    # Free all the resource objects of the pool
    pool.free_pool()


if __name__ == "__main__":
    main()


"""
Output:

Creating new resource...
https://www.dataresources.com/server_entry_point.php
Creating new resource...
https://www.another-dataresources.com/another_server_entry_point.php
Reusing existing resource...

Reusing existing resource...

"""
